internal/cli/base.go

```diff
@@ -232,114 +236,55 @@ func (c *baseCommand) Init(opts ...Option) error {
-				// other arguments
-			} else if !baseCfg.ProjectTargetRequired {
-				// Assume the target is just project
-				p := c.args[0]
-				c.refProject = &pb.Ref_Project{Project: p}
-				// We don't explicitly set the app because there was none requested,
-				// and we might or might not be working on an app later.
-
-				// Shift the args
-				c.args = c.args[1:]
-
-				// Explicitly set remote
-				c.flagRemote = true
+		// If that worked, set our refs
+		if cfg != nil {
+			c.refProject = &pb.Ref_Project{Project: cfg.Project}
+			for _, app := range cfg.Apps() {
+				c.refApps = append(c.refApps, &pb.Ref_Application{
+					Project:     cfg.Project,
+					Application: app,
+				})
 			}
 		}
+	}
 
-		// If we didn't get our ref, then we need to load config
-		if c.refApp == nil && c.refProject == nil {
-			// We look at both app and project for the case where no target was specified
-			// i.e. already in a project directory
-			baseCfg.Config = true
-		}
+	// Parse project flags
+	if c.flagProject != "" {
+		c.refProject = &pb.Ref_Project{Project: c.flagProject}
 	}
 
-	// If we're loading the config, then get it.
-	if baseCfg.Config {
-		cfg, err := c.initConfig("", baseCfg.ConfigOptional)
-		if err != nil {
-			c.logError(c.Log, "failed to load config", err)
-			return err
-		}
+	if baseCfg.ProjectTargetRequired && c.refProject == nil {
+		// The user must not have specified a project flag, and config parsing didn't produce one either.
 
-		c.cfg = cfg
-		if cfg != nil {
-			project := &pb.Ref_Project{Project: cfg.Project}
+		// NOTE(izaak) The UX here will be refined in the next pass - it's ok that this is terse for now.
+		err := errors.New("No project specified, and no waypoint.hcl found.")
+		c.ui.Output(clierrors.Humanize(err), terminal.WithErrorStyle())
+		return err
+	}
 
-			// If we're loading config, we'll have a project, and we set it now.
-			// If they didn't provide a value via flag, we default to
-			// the project from initConfig.
-			if c.flagProject != "" {
-				project = &pb.Ref_Project{Project: c.flagProject}
-			}
-			if c.refProject == nil {
-				c.refProject = project
-			}
+	// Parse app flags
+	if c.flagApp != "" {
+		// NOTE: we could allow app to be specified multiple times in the future
+		c.refApps = []*pb.Ref_Application{{Application: c.flagApp}}
+	}
 
-			// If we require an app target and we still haven't set it,
-			// and the user provided it via the CLI, set it now. This code
-			// path is only reached if it wasn't set via the args either
-			// above.
-			if baseCfg.AppTargetRequired &&
-				c.refApp == nil &&
-				c.flagApp != "" {
-				c.refApp = &pb.Ref_Application{
-					Project:     project.Project,
-					Application: c.flagApp,
-				}
-			}
-		}
+	if (baseCfg.SingleAppTarget || baseCfg.MultiAppTarget) && len(c.refApps) == 0 {
+		err = fmt.Errorf("This command requires an app to be targeted, but no apps were found in project %q.", c.refProject.Project)
```

Comment:
Does this miss the flow where the user specifies a `-project` for a multi-app-target command, we don't have a waypoint.hcl to load apps from, and we haven't yet collected all the apps assigned to the specified project after setting the Project on line 260?

Reply:
Good catch! Fixing. We'll need to make an api call to list the project here to fix it.